Add commands tests

eps90 opened the pull request:
Hey there!
This PR introduces tests of few commands including:
* `AbstractCommand`
* `CompileCommand`
* `ListCommand`
* `LockCommand`
* `UnlockCommand`

These tests cover the code in 100% (line coverage) without changing current functionality.
Few words about PR:
* I used [malkusch\php-mock](https://github.com/malkusch/php-mock) for creating mocks of PHP's built-in functions. As far as I know, it uses the fact that PHP interpreter will use the function with the same name from the namespace where it was used in. This library contains some other cool options like making **stubs** on built-in functions but since we're keeping compatibility with PHP 5.3, we can't use them and we can just mock them to avoid unexpected results. Fortunatelly, author created the `php-5.3` branch.
* I had to use `@uses` annotation to point that we use some other classes (like statics, impossible to inject and then mock) and not cover them. Without that annotation PHPUnit run with `--coverage-*` trigger will mark all tests as risky. **Note:** I didn't use this annotation only to suppress warnings. Each test class was tested without them to see if there are no other classes under control. As you can see, each test class contains almost the same `@uses` annotations.
* I had to add setter in `CompileCommand` to mock `Mage\Compiler` class. This setter is called in (also created by me) constructor, so nothing changed in this class' functionality.

For now coverage of the project is **22.36%**. Before creating these tests, it was about 7%.
Feel free to review and discuss! Thanks!

SenseException commented:
If I can make it possible, I'll take a look at it over the weekend. It is too bad that Reflection is needed, but maybe we'll get some inspiration for other ideas for this ticket and #175 sometime.

eps90 commented:
@SenseException Yes, I know it's ugly that I used Reflection. This is the way I test value objects. I didn't want to set it by setting by setter and get by getter. I've never found any other way to do this.
In the other hand, I could omit testing those methods since their complexity is 1.. and I do simple assignment there. So I can omit them or disable coverage there ;) Especially it's just an abstract class.
Thanks for feedback! ;)

SenseException commented on Mage/Command/BuiltIn/CompileCommand.php:
```diff
@@ -22,19 +22,43 @@
+    {
+        $this->setCompiler(new Compiler());
+    }
+
```
Why not the following constructor code?

``` php
    /**
     * @var Compiler
     */
    private $compiler;

    public function __construct(Compiler $compiler = null)
    {
        if (is_null($compiler)) {
            $compiler = new Compiler();
        }
        
        $this->compiler = $compiler;
    }
```

This way, you don't need the setter and won't allow to change the compiler on runtime like the previews version.

SenseException commented on Mage/Command/BuiltIn/CompileCommand.php:
```diff
@@ -22,19 +22,43 @@
+        $this->setCompiler(new Compiler());
+    }
+
+    public function setCompiler(Compiler $compiler)
```
If you want to keep the setter and the possibility to set the compiler on runtime, the doc-block is missing for this method.

eps90 replied:
Why?

SenseException replied:
For the different IDE's that are using it. Some may need it and it may help other contributors.

And maybe someday we create a API documentation for the project :-)

eps90 replied:
API is ready but it still need improvements. It should be updated per deploy. I think it's subject for another issue :)
I don't know how other IDEs behave, but in PHPStorm it's enough to give static-type of method's param. If it's what you're talking about.

SenseException commented on composer.json:
```diff
@@ -9,13 +9,15 @@
             "Task\\": [".mage/tasks", "../../../.mage/tasks"],
-            "Command\\": [".mage/tasks", "../../../.mage/commands"]
+            "Command\\": [".mage/tasks", "../../../.mage/commands"],
+            "MageTest\\": "./tests/MageTest"
```
For unit test classes, please use autoload-dev.
https://getcomposer.org/doc/04-schema.md#autoload-dev

eps90 replied:
good point.

SenseException commented on tests/MageTest/Command/AbstractCommandTest.php:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * @covers ::getConfig
+     */
+    public function testGetConfig()
```
It is considered for some people as not a good practice, but what do you think about the depends-annotation? Maybe not a separated unit running for itself, but a possibility not to use Reflection for the getter test.

eps90 replied:
I agree, I wanted just test the current code with almost no changes. I also agree that dependency injection would help us a lot but for now I wanted only to test if the classes are constructed properly, nothing else. i think we shouldn't do any refactor till we test the weakest points of the project.

SenseException replied:
There is no refactoring needed for using depends-annotation. When used, one of the tests doesn't need Reflection anymore but depends on the positive test result of the other one. I think the less we are using Reflection functionality together with the tests, the more stable it gets, because we don't "test" Reflection with it.

eps90 replied:
Umm ok. Sorry for my yesterday note, I was thinking of dependency injection instead of depend-annotation (I was exhausted, don't ask why).
I see advantages of such solution, but some testing philosophies say that tests should be completely independent. In the other hand these are just setters, the methods with path complexity equal 1. So well.. I need to consider it. But thanks for spotting! ;)

SenseException commented on tests/MageTest/Command/AbstractCommandTest.php:
```diff
@@ -0,0 +1,52 @@
+        $this->abstractCommand->setConfig($configMock);
+
+        $actual = $this->getPropertyValue($this->abstractCommand, 'config');
+        $this->assertEquals($configMock, $actual);
```
You actually expect the same instance you set with the setter, not another config instance. You need to use

``` php
$this->assertSame($configMock, $actual);
```

SenseException commented on tests/MageTest/Command/AbstractCommandTest.php:
```diff
@@ -0,0 +1,52 @@
+        $this->setPropertyValue($this->abstractCommand, 'config', $configMock);
+
+        $actual = $this->abstractCommand->getConfig();
+        $this->assertEquals($configMock, $actual);
```
Same as mentioned in your setter test. You need `assertSame`:

``` php
$this->assertSame($configMock, $actual);
```

SenseException commented on tests/MageTest/Command/BuiltIn/CompileCommandTest.php:
```diff
@@ -0,0 +1,114 @@
+     * @covers ::__construct
+     * @covers ::setCompiler
+     */
+    public function testConstruct()
```
Does a constructor needs an own test when it only sets a property?

SenseException commented on tests/MageTest/Command/BuiltIn/CompileCommandTest.php:
```diff
@@ -0,0 +1,114 @@
+     * @covers ::__construct
+     * @covers ::setCompiler
+     */
+    public function testSetCompiler()
```
See my comment in Compiler class part about Setter.

eps90 commented:
Thanks @SenseException . I look at your comments in free time. Now I'm quite busy, but I think tomorrow evening I'll do some fixes into my PR. Thanks a lot! ;)

SenseException commented on tests/MageTest/Command/BuiltIn/UnlockCommandTest.php:
```diff
@@ -0,0 +1,125 @@
+        $this->assertEquals($expectedExitCode, $actualExitCode);
+        $this->assertEquals($expectedLockFilePath, self::$fileExistsResult);
+    }
+}
```
Both tests,testRun and testRunIfFileNotExist, are sharing nearly the same code. Maybe this can be excluded into a common used method.

eps90 replied:
cetrainly :)

SenseException commented:
@eps90 I see you already noticed my first comments on some code :-)
Good work adding tests for those commands.

The malkusch/php-mock project is a nice help, but we should agree on, that the goal is to work toward a code, that doesn't need this anymore in the future. We may mock a behaviour away, that can differ between PHP Versions without knowing it. I don't know if we manage to remove it completly one day, but maybe it is possible. For now, we should use it. Filesystem access can someday be moved into an own Filesystem class, that can be tested separately with vfsStream.

eps90 commented:
I totally agree with wyou, @SenseException . But as I said before, I didn't want to change class' contents till we have a test complet. I think if all commands will be done (tested), we can think about refactor. In `ListCommand` there's even a fatal exception if you try to run this command without initalizing the Magallanes. I could write the tests for it, but it's topic for another PR. That's why I used `php-mock` to mock built-in functions. In the future it's obvious to have `Filesystem` class :) Little steps and everything will be shining! :)

eps90 commented:
And sorry again, I won't handle to make fixes today.

SenseException commented:
@eps90 We do this in out free time so there's no reason to apologize.

eps90 commented:
@SenseException I fixed most of things you note. I've left only that `@depends` annotation till we find out is it really a good practice (it's not in my opinion, but it has own advantages). Instead, I moved testing getters and setters to `BaseTest` class.
Thanks for reviewing!

eps90 commented:
@SenseException ... and I think this could be merged, because you're the only person here which I discuss about unit tests here ;)

eps90 commented:
@SenseException fixed your notes. You can re-review it :)

SenseException commented:
edpauto/Magallanes@cf0f15f4ce59fbcffea02d07ef41683f5ee1495c is handled with edpauto/Magallanes@832435e07ceef70d4aa7d38f7e3e5cfeb33863d8 
edpauto/Magallanes@f8804182f5f07f86babed2b902795310190cfcfa is handled in edpauto/Magallanes@3f02ebcfe4a2878d46103d7e19ee11a2b756675e

It's ok now.

eps90 commented:
Great, thanks a lot!
